### PR TITLE
Add ingester (and kafka) support

### PR DESCRIPTION
**Pull request by objectiser**
This is an initial suggestion for how we can easily switch from a simple production deployment (e.g. collector and query using a common backend persistent storage) to one where Kafka is used as a buffer between collectors (publishing to Kafka) and ingesters (consuming from Kafka and storing in the specified storage).

The difference between the two configurations would be:
```
apiVersion: io.jaegertracing/v1alpha1
kind: Jaeger
metadata:
  name: simple-prod
spec:
  strategy: production
  storage:
    type: elasticsearch
    options:
      es:
        server-urls: http://elasticsearch:9200
```
and
```
apiVersion: io.jaegertracing/v1alpha1
kind: Jaeger
metadata:
  name: simple-prod
spec:
  strategy: production
  ingester:
    options:
      kafka:
        brokers: 127.0.0.1:9092
        topic: jaeger-spans
  storage:
    type: elasticsearch
    options:
      es:
        server-urls: http://elasticsearch:9200
```

This is a simple change to the configuration, i.e. defining the ingester options, but not necessarily clear this results in a more complex architecture.

Issues to consider:
* Both the collector and ingester would have `storage` definitions - whereas the CR only has one area to define these. We could list storage types, e.g. `type: elasticsearch, kafka` with options for both, but it is still not clear which should be used in collector and ingester.
* We need an `ingester` node in the CR, equivalent to the other top level nodes, to be able to override the image - but these are just used to define information about the individual components. Not necessarily how those components collaborate with each other.

So the question is whether defining the ingester should imply using the `collector->kafka->ingester` approach and that the storage option therefore moves to the final component in the chain - i.e. ingester rather than collector.

If the existence of ingester options is not considered appropriate, then I would recommend we define more explicit strategies:
* `allInOne` as now, single deployment with agent, collector and query
* collector and query with shared storage
* collector->kafka->ingester and query with shared storage

(what we name the 2nd and 3rd strategies is another matter)

I think this is actually the best approach, as it is clear what architecture the user wants, and the CR can then be verified to ensure it has all of the required information defined for each of the components. It also means that the `ingester` node is optional - as the kafka options could just be listed in the `storage.options`, but would still result in the ingester being configured.

Feedback on the current PR and proposed change would be appreciated - however I might not work on it much until new year.

(Also requires e2e tests)

Signed-off-by: Gary Brown <gary@brownuk.com>

### CODE REVIEW

**Comment by jpkrohling** [comment]
This change is [Reviewable](https://reviewable.io/reviews/jaegertracing/jaeger-operator/168)

File: pkg/deployment/collector.go [hunk 1]
```diff
@@ -64,8 +64,15 @@ func (c *Collector) Get() *appsv1.Deployment {
 		})
 	}
 
+	storageType := c.jaeger.Spec.Storage.Type
+	// If ingester options have been defined, then change storage type
+	// to Kafka, and the storage options will be used in the Ingester instead
+	if len(c.jaeger.Spec.Ingester.Options.Map()) > 0 {
```
Comment (pavolloffay) [inline pkg/deployment/collector.go]: This is interesting, do we assume ingester options have to be provided? 

Or in other words are not we supporting the case where no conf props are needed?

File: pkg/deployment/collector_test.go [hunk 2]
```diff
@@ -266,3 +267,74 @@ func TestCollectorResources(t *testing.T) {
...
+					"es.server-urls": "http://somewhere",
+				}),
+			},
+		},
+	}
+	collector := NewCollector(jaeger)
+	dep := collector.Get()
+
+	envvars := []v1.EnvVar{
+		v1.EnvVar{
```
Comment (pavolloffay) [inline pkg/deployment/collector_test.go]: nit: types can be omitted here

File: pkg/deployment/ingester.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+)
+
+// Ingester builds pods for jaegertracing/jaeger-ingester
+type Ingester struct {
+	jaeger *v1alpha1.Jaeger
+}
+
+// NewIngester builds a new Ingester struct based on the given spec
+func NewIngester(jaeger *v1alpha1.Jaeger) *Ingester {
+	if jaeger.Spec.Ingester.Size <= 0 {
```
Comment (pavolloffay) [inline pkg/deployment/ingester.go]: can we move this to normalize func?

File: pkg/deployment/ingester.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+	if jaeger.Spec.Ingester.Image == "" {
+		jaeger.Spec.Ingester.Image = fmt.Sprintf("%s:%s", viper.GetString("jaeger-ingester-image"), viper.GetString("jaeger-version"))
+	}
+
+	return &Ingester{jaeger: jaeger}
+}
+
+// Get returns a ingester pod
+func (i *Ingester) Get() *appsv1.Deployment {
```
Comment (pavolloffay) [inline pkg/deployment/ingester.go]: Thinking about consistency here...

The other creators in this repo returns object not pointer. Maybe a better to standardize and include func which check if the ingester is needed

File: pkg/deployment/collector_test.go (hunk 2 above)
Comment (pavolloffay) [reply pkg/deployment/collector_test.go]: applies to other places in this repo as well

File: pkg/deployment/collector.go (hunk 1 above)
Comment (objectiser) [reply pkg/deployment/collector.go]: This was the easiest approach, as I believe ingester options will always be required - but is not intuitive. As mentioned in the PR description, I think a better approach would be to have more distinct strategies that effectively define the architectural structure being used. So all-in-one, another for the current production setup, and a new third option used to represent the kafka based approach.

File: pkg/deployment/ingester.go (hunk 3 above)
Comment (objectiser) [reply pkg/deployment/ingester.go]: Possibly, but currently this is consistent with the other components. So should be handled in one go in another PR.

File: pkg/deployment/ingester.go (hunk 4 above)
Comment (objectiser) [reply pkg/deployment/ingester.go]: Possibly - but the other deployments in this package (query, collector, agent) all return pointer. So this is consistent with the other deployments at the moment - so if going to be changed, it should be done in separate PR for all components.

**Comment by objectiser** [comment]
This PR is on hold until @jpkrohling returns so we can resolve how the operator should know when ingester is being used.

**Comment by jpkrohling** [comment]
I trust your judgment about which approach is better. Initially, I would say that having the ingester options are not a problem, but I agree that it's cleaner to have the three options as you listed them. The main motivation as I see it is that we could implement new features specific for the strategy in a way that is transparent to the user. This means that we are making the operator *more* opinionated, which I see as a good thing.

About the naming, how about:

* `allInOne`
* `production`
* `productionHA` (highly-available), implying that it's `production + other stuff`

**Comment by objectiser** [comment]
@jpkrohling I was hoping we could move away from `production` (so deprecate it) and use two more appropriate names, as `productionHA` doesn't really describe the kafka based architecture, as the `production` strategy can also be HA.

So I think we are agreed that the strategy name is required for kafka/ingester - but just need to resolve the actual names to use.

**Comment by objectiser** [comment]
@jpkrohling Created a new strategy called `streaming` which seems appropriate - so we could leave `production` for now, although in the future we may want to consider a rename, as it implies it is the only _production ready_ architecture.

If this approach seems reasonable, then next step would be to add some e2e tests and docs.

**Comment by jpkrohling** [comment]
+1 from my side!

File: README.adoc [hunk 5]
```diff
@@ -150,6 +150,52 @@ spec:
...
+
+The main backend components, agent, collector and query service, are all packaged into a single executable which is configured (by default) to use in-memory storage.
+
+=== Production
+
+The `production` strategy is intended (as the name suggests) for production environments, where long term storage of trace data is important, as well as a more scalable and highly available architecture is required. Each of the backend components is therefore separately deployed.
+
+The agent can be injected as a sidecar on the instrumented application or as a daemonset.
+
+The query and collector services are configured with a supported storage type - currently cassandra or elasticsearch. Multiple instances of each of these components can be provisions as required for performance and resilience purposes.
```
Comment (jpkrohling) [inline README.adoc]: `s/can be provisions/can be provisioned/`?

File: README.adoc [hunk 6]
```diff
@@ -150,6 +150,52 @@ spec:
...
+    storage:
+      type: elasticsearch
+      options:
+        es:
+          server-urls: http://elasticsearch:9200
+----
+
+=== Streaming
+
+The `streaming` strategy is designed to augment the `production` strategy by providing a streaming capability that effectively sits between the collector and the backend storage (e.g. cassandra or elasticaearch). This provides the benefit of reducing the pressure on the backend storage, under high load situations, and enables other trace post processing capabilities to tap into the real time span data directly from the streaming platform (kafka).
```
Comment (jpkrohling) [inline README.adoc]: `s/elasticaearch/elasticsearch/`

Comment (objectiser) [reply README.adoc]: sorry should have re-read :)

**Comment by objectiser** [comment]
@pavolloffay don't think any of your remaining points were blockers - but can be dealt with in separate PR if they are still issues.

**Pull request merged**